File: app/src/main/java/com/nominacopro/domain/payperiod/PayPeriodCalculator.kt
package com.nominacopro.domain.payperiod

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.TextStyle
import java.util.Locale

enum class PayPeriodType(val label: String) {
    MONTHLY("Mensual"),
    WEEKLY("Semanal"),
    BIWEEKLY("Quincenal"),
    VENTEEN("Ventena (21 días)"),
    ;

    val hasSubPeriods: Boolean get() = this != MONTHLY

    companion object {
        fun fromStored(value: String?): PayPeriodType =
            entries.find { it.name == value } ?: BIWEEKLY
    }
}

data class PayPeriod(
    val start: LocalDate,
    val end: LocalDate,
    val label: String,
    val indexInMonth: Int = 0,
) {
    val dayCount: Int =
        java.time.temporal.ChronoUnit.DAYS.between(start, end).toInt() + 1
}

object PayPeriodCalculator {
    private val locale = Locale("es", "CO")

    fun periodsInMonth(type: PayPeriodType, yearMonth: YearMonth): List<PayPeriod> =
        when (type) {
            PayPeriodType.MONTHLY -> emptyList()
            PayPeriodType.WEEKLY -> weeklyPeriods(yearMonth)
            PayPeriodType.BIWEEKLY -> biweeklyPeriods(yearMonth)
            PayPeriodType.VENTEEN -> venteenPeriods(yearMonth)
        }

    fun defaultPeriodIndex(type: PayPeriodType, yearMonth: YearMonth, today: LocalDate = LocalDate.now()): Int {
        if (!type.hasSubPeriods) return 0
        val periods = periodsInMonth(type, yearMonth)
        if (periods.isEmpty()) return 0
        if (yearMonth != YearMonth.from(today)) return 0
        return periods.indexOfFirst { today in it.start..it.end }.coerceAtLeast(0)
    }

    private fun biweeklyPeriods(ym: YearMonth): List<PayPeriod> {
        val monthLabel = ym.month.getDisplayName(TextStyle.SHORT, locale)
        val lastDay = ym.lengthOfMonth()
        return buildList {
            add(period(ym.atDay(1), ym.atDay(minOf(15, lastDay)), "1–${minOf(15, lastDay)} $monthLabel", 0))
            if (lastDay >= 16) {
                add(period(ym.atDay(16), ym.atEndOfMonth(), "16–$lastDay $monthLabel", 1))
            }
        }
    }

    private fun weeklyPeriods(ym: YearMonth): List<PayPeriod> {
        val monthStart = ym.atDay(1)
        val monthEnd = ym.atEndOfMonth()
        val monthLabel = ym.month.getDisplayName(TextStyle.SHORT, locale)
        var weekStart = monthStart.with(java.time.temporal.TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        val periods = mutableListOf<PayPeriod>()
        while (weekStart <= monthEnd) {
            val weekEnd = weekStart.plusDays(6)
            val overlapStart = maxOf(weekStart, monthStart)
            val overlapEnd = minOf(weekEnd, monthEnd)
            if (overlapStart <= overlapEnd) {
                periods.add(
                    period(
                        weekStart,
                        weekEnd,
                        "Sem ${periods.size + 1} · ${overlapStart.dayOfMonth}–${overlapEnd.dayOfMonth} $monthLabel",
                        periods.size,
                    ),
                )
            }
            weekStart = weekStart.plusWeeks(1)
        }
        return periods
    }

    private fun venteenPeriods(ym: YearMonth): List<PayPeriod> {
        val monthStart = ym.atDay(1)
        val monthEnd = ym.atEndOfMonth()
        val monthLabel = ym.month.getDisplayName(TextStyle.SHORT, locale)
        var periodStart = LocalDate.of(ym.year, 1, 1)
        val periods = mutableListOf<PayPeriod>()
        var ventenaIndex = 1
        while (periodStart <= monthEnd && periodStart.year == ym.year) {
            val periodEnd = periodStart.plusDays(20)
            if (periodEnd >= monthStart && periodStart <= monthEnd) {
                val overlapStart = maxOf(periodStart, monthStart)
                val overlapEnd = minOf(periodEnd, monthEnd)
                periods.add(
                    period(
                        periodStart,
                        periodEnd,
                        "Ventena $ventenaIndex · ${overlapStart.dayOfMonth}–${overlapEnd.dayOfMonth} $monthLabel",
                        periods.size,
                    ),
                )
                ventenaIndex++
            }
            periodStart = periodStart.plusDays(21)
        }
        return periods
    }

    private fun period(start: LocalDate, end: LocalDate, label: String, index: Int) =
        PayPeriod(start = start, end = end, label = label, indexInMonth = index)
}

File: web/lib/payroll/payPeriodCalculator.ts
import { formatMonthName } from "@/lib/format";
import {
  addDays,
  addWeeks,
  endOfMonth,
  isAfter,
  isBefore,
  parseYearMonth,
  previousOrSameMonday,
  startOfMonth,
  type LocalDate,
} from "./dates";
import type { PayPeriodType } from "./models";

export interface PayPeriod {
  start: LocalDate;
  end: LocalDate;
  label: string;
  indexInMonth: number;
}

export function payPeriodDayCount(period: PayPeriod): number {
  const start = parseDateParts(period.start);
  const end = parseDateParts(period.end);
  const ms = Date.UTC(end.y, end.m - 1, end.d) - Date.UTC(start.y, start.m - 1, start.d);
  return Math.floor(ms / 86400000) + 1;
}

function parseDateParts(iso: LocalDate) {
  const [y, m, d] = iso.split("-").map(Number);
  return { y, m, d };
}

export function hasSubPeriods(type: PayPeriodType): boolean {
  return type !== "MONTHLY";
}

export function shouldShowSubPeriods(type: PayPeriodType, periods: PayPeriod[]): boolean {
  return hasSubPeriods(type) && periods.length > 0;
}

export function periodsInMonth(type: PayPeriodType, yearMonth: string): PayPeriod[] {
  const { year, month } = parseYearMonth(yearMonth);
  switch (type) {
    case "MONTHLY":
      return [];
    case "WEEKLY":
      return weeklyPeriods(year, month);
    case "BIWEEKLY":
      return biweeklyPeriods(year, month);
    case "VENTEEN":
      return venteenPeriods(year, month);
    default:
      return biweeklyPeriods(year, month);
  }
}

export function defaultPeriodIndex(
  type: PayPeriodType,
  yearMonth: string,
  today: LocalDate = todayIso(),
): number {
  if (!hasSubPeriods(type)) return 0;
  const periods = periodsInMonth(type, yearMonth);
  if (periods.length === 0) return 0;
  const now = new Date();
  const currentYm = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
  if (yearMonth !== currentYm) return 0;
  const idx = periods.findIndex((p) => !isBefore(today, p.start) && !isAfter(today, p.end));
  return Math.max(0, idx);
}

function todayIso(): LocalDate {
  const now = new Date();
  return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")}`;
}

function biweeklyPeriods(year: number, month: number): PayPeriod[] {
  const monthLabel = formatMonthName(month);
  const lastDay = new Date(year, month, 0).getDate();
  const periods: PayPeriod[] = [];
  const firstEnd = Math.min(15, lastDay);
  periods.push(
    period(startOfMonth(year, month), `${year}-${String(month).padStart(2, "0")}-${String(firstEnd).padStart(2, "0")}`, `1–${firstEnd} ${monthLabel}`, 0),
  );
  if (lastDay >= 16) {
    periods.push(
      period(
        `${year}-${String(month).padStart(2, "0")}-16`,
        endOfMonth(year, month),
        `16–${lastDay} ${monthLabel}`,
        1,
      ),
    );
  }
  return periods;
}

function weeklyPeriods(year: number, month: number): PayPeriod[] {
  const monthStart = startOfMonth(year, month);
  const monthEnd = endOfMonth(year, month);
  const monthLabel = formatMonthName(month);
  let weekStart = previousOrSameMonday(monthStart);
  const periods: PayPeriod[] = [];
  while (!isAfter(weekStart, monthEnd)) {
    const weekEnd = addDays(weekStart, 6);
    const overlapStart = isBefore(weekStart, monthStart) ? monthStart : weekStart;
    const overlapEnd = isAfter(weekEnd, monthEnd) ? monthEnd : weekEnd;
    if (!isAfter(overlapStart, overlapEnd)) {
      const startDay = parseDateParts(overlapStart).d;
      const endDay = parseDateParts(overlapEnd).d;
      periods.push(
        period(
          weekStart,
          weekEnd,
          `Sem ${periods.length + 1} · ${startDay}–${endDay} ${monthLabel}`,
          periods.length,
        ),
      );
    }
    weekStart = addWeeks(weekStart, 1);
  }
  return periods;
}

function venteenPeriods(year: number, month: number): PayPeriod[] {
  const monthStart = startOfMonth(year, month);
  const monthEnd = endOfMonth(year, month);
  const monthLabel = formatMonthName(month);
  let periodStart: LocalDate = `${year}-01-01`;
  const periods: PayPeriod[] = [];
  let ventenaIndex = 1;
  while (!isAfter(periodStart, monthEnd) && periodStart.startsWith(String(year))) {
    const periodEnd = addDays(periodStart, 20);
    if (!isBefore(periodEnd, monthStart) && !isAfter(periodStart, monthEnd)) {
      const overlapStart = isBefore(periodStart, monthStart) ? monthStart : periodStart;
      const overlapEnd = isAfter(periodEnd, monthEnd) ? monthEnd : periodEnd;
      const startDay = parseDateParts(overlapStart).d;
      const endDay = parseDateParts(overlapEnd).d;
      periods.push(
        period(
          periodStart,
          periodEnd,
          `Ventena ${ventenaIndex} · ${startDay}–${endDay} ${monthLabel}`,
          periods.length,
        ),
      );
      ventenaIndex++;
    }
    periodStart = addDays(periodStart, 21);
  }
  return periods;
}

function period(start: LocalDate, end: LocalDate, label: string, index: number): PayPeriod {
  return { start, end, label, indexInMonth: index };
}
